lib/seahorn/BvOpSem2.cc

```diff
@@ -3182,6 +3186,45 @@ Optional<APInt> Bv2OpSem::agg(Type *aggTy,
   return res;
 }
 
+/// Crab stuff
+// Build Crab CFG, run crab analyses, etc.
+void Bv2OpSem::mkCrabAnalysis(const llvm::Module &M,
+                              seadsa::GlobalAnalysis &dsa,
+                              TargetLibraryInfoWrapperPass &tli) {
+
+  std::unique_ptr<clam::HeapAbstraction> heap_abs(
+      new clam::SeaDsaHeapAbstraction(M, dsa));
+
+  // -- Set parameters for CFG
+  clam::CrabBuilderParams cfg_builder_params;
+  LOG("bmc-crab", cfg_builder_params.print_cfg = true;);
+  // Note: default precision is NUM
+  cfg_builder_params.interprocedural = true;
+
+  m_cfg_builder_man.reset(new clam::CrabBuilderManager(cfg_builder_params, tli,
+                                                       std::move(heap_abs)));
+
+  // -- Initialize crab for solving ranges
+  m_crab_rng_solver.reset(new clam::InterGlobalClam(M, *m_cfg_builder_man));
+
+  /// Set Crab parameters
+  clam::AnalysisParams aparams;
+  aparams.dom = clam::CrabDomain::INTERVALS;
+  aparams.run_inter = true;
+  aparams.check = clam::CheckerKind::NOCHECKS;
+  /// Run the Crab analysis
+  clam::ClamGlobalAnalysis::abs_dom_map_t assumptions;
+  // aparams.print_invars = true;
+  m_crab_rng_solver->analyze(aparams, assumptions);
+}
+
+const clam::ClamQueryAPI::Range
+Bv2OpSem::getInstRng(const Instruction &I) const {
+  if (!m_crab_rng_solver)
```

Comment:
llvm also provides ranges for variables. Use that in addition to what crab provides. Use that if crab does not provide.

I don't know what the type of the output is, but returning `{0, 0}` seems wrong since this is a concrete range.

Reply:
@yusen: please update Clam. I just pushed a commit so that we use `llvm::ConstantRange` instead of `clam::ClamQueryAPI::Range`

Reply:
Ok. I will use an empty constant range to represent **no analysis provided**.

Reply:
For that, `getFull` represents top (https://llvm.org/doxygen/classllvm_1_1ConstantRange.html#a20bc7c6f540132189d07ad7f73bda1ee).
Btw, clam generates all the ConstantRange's with 64 bits. This is needed because ConstantRange are intervals that supports wraparound so the bitwidth is important. Make sure if you generate ConstantRange's and compare with those generate by Clam because they probably need to have the same bitwidth.

Reply:
Ok. Thanks for the advice.